nrfx_config_nrf9160: Remove dependency on DT from NRF_* mappings

anangl opened the pull request:
Stop using base addresses of peripherals extracted from devicetree
in definitions of NRF_* symbols that are needed by HALs and drivers.
Instead, in order to map to secure or non-secure base addresses of
peripherals, take corresponding symbols provided by MDK (with _S or
_NS suffix) depending on the Kconfig option ARM_NONSECURE_FIRMWARE.

This commit provides also temporary mappings of the GPIOTE IRQ number
making it possible to use nrfx_gpiote driver on nRF9160. These mappings
are supposed to be removed when the nrfx_gpiote driver is corrected to
not use the GPIOTE_IRQn symbol.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
@kl-cruz Could you take a look as well?

carlescufi commented on nrfx_config_nrf9160.h:
```diff
@@ -36,243 +36,124 @@
-#define NRF_RTC1 \
-        ((NRF_RTC_Type *)DT_NORDIC_NRF_RTC_RTC_1_BASE_ADDRESS)
-#endif
+#if defined(CONFIG_ARM_NONSECURE_FIRMWARE)
```
You could've written a macro:

```
#define NRF_PERIPH(P) #if defined(CONFIG_ARM_NONSECURE_FIRMWARE) \
                                            #define P P##_NS \
                                      #else \
                                           #define P P##_S \
                                      #endif

NRF_PERIPH(NRF_CLOCK);
NRF_PERIPH(NRF_DPPIC);
...
```
That would avoid duplication of the lists.
What do you think?

anangl replied:
Yes, I think it could work.

We've got one exception: `NRF_GPIOTE` that needs to be mapped to `NRF_GPIOTE1_NS` or `NRF_GPIOTE0_S`, but it can be handled separately (it will be more visible then, what is good).

Another thing is that some peripherals are accessible only from secure code, so only `*_S` symbols are defined for them. But this also should not be a problem. The only impact I can see is that it would not be possible to use something like `#if defined(NRF_CRYPTOCELL)` to check if the CRYPTOCELL peripheral is available, because `NRF_CRYPTOCELL` will be defined also for non-secure firmware. But we don't do such checks.

I agree to apply such modification. Any objections from others?

carlescufi replied:
Yep, just use special casing for those that are not symmetric in the secure vs non-secure worlds.

ioannisg commented:
I understand the change, but I would like a summary of why this is needed...

kl-cruz commented:
> @kl-cruz Could you take a look as well?

ofc!

carlescufi commented:
> I understand the change, but I would like a summary of why this is needed...

It is not needed, but I really don't see the benefit of using DT in the context of the HAL/nrfx, since that already comes with macros that define the peripherals.

anangl commented:
> > I understand the change, but I would like a summary of why this is needed...
> 
> It is not needed, but I really don't see the benefit of using DT in the context of the HAL/nrfx, since that already comes with macros that define the peripherals.

Yes, it is not needed, but it will make our and users' life easier. Maintenance of these mappings in their current form is quite troublesome (for instance, we'll need to introduce the "nordic,nrf-spim" compatible, to differentiate nodes in DT, like in the case of "nordic,nrf-uart" and "nordic,nrf-uarte", and such change will affect these mappings, although it is not really related). Users are also forced to enable a given node in devicetree just to be able to use a given NRF_* symbol, which is not really nice (see [spm](https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/master/subsys/spm/spm.c#L339) as an example).

anangl commented:
I used the macro proposed by @carlescufi.

ioannisg commented:
> 
> 
> > I understand the change, but I would like a summary of why this is needed...
> 
> It is not needed, but I really don't see the benefit of using DT in the context of the HAL/nrfx, since that already comes with macros that define the peripherals.

The problem, @carlescufi @anangl , will be, if we start using the DT_* macros anywhere else in the code? How can we exclude that?

If we do use them, we need to be making sure that they match the NRF_ macros we define here. Do we need build-time asserts for that?

carlescufi commented:
> > > I understand the change, but I would like a summary of why this is needed...
> > 
> > 
> > It is not needed, but I really don't see the benefit of using DT in the context of the HAL/nrfx, since that already comes with macros that define the peripherals.
> 
> The problem, @carlescufi @anangl , will be, if we start using the DT_* macros anywhere else in the code. Do we do that? How can we exclude that?
> 
> If we do use them, we need to be making sure that they match the NRF_ macros we define here. Do we need build-time asserts for that?

This is to configure nrfx, not Zephyr. nrfx comes with a hardcoded list of peripherals and their addresses, and I think it makes sense to use those. If we ever have a mismatch between DT and nrfx then it will clearly show as the peripheral not working. I am in favor thus of using nrfx macros to configure nrfx and DT to configure the Zephyr shim layers.

ioannisg commented on nrfx_config_nrf9160.h:
```diff
@@ -36,243 +36,83 @@
+#define NRF_WDT         NRF_PERIPH(NRF_WDT)
+
+
+#if defined(CONFIG_ARM_NONSECURE_FIRMWARE)
```
This part about GPIOTE would not be needed, if we used the DTS-way to define these. Because we have already defined different GPIOTEs in DTS for the two S and NS board versions

carlescufi replied:
I agree, but the overall point still stands. This is the "nrfx way"of configuring nrfx, and I think it's more consistent with other nrfx uses.

ioannisg replied:
> This is to configure nrfx, not Zephyr. nrfx comes with a hardcoded list of peripherals and their addresses, and I think it makes sense to use those. If we ever have a mismatch between DT and nrfx then it will clearly show as the peripheral not working. I am in favor thus of using nrfx macros to configure nrfx and DT to configure the Zephyr shim layers.

But Zephyr drivers, now depend on this, to get the peripheral base addresses. So it seems to me that nRF drivers in Zephyr get #define macros from both nRFx and DTS in the source code. Can we clean this up?

ioannisg commented on nrfx_config_nrf9160.h:
```diff
@@ -36,243 +36,83 @@
-#define NRF_RTC1 \
-        ((NRF_RTC_Type *)DT_NORDIC_NRF_RTC_RTC_1_BASE_ADDRESS)
-#endif
+#if defined(CONFIG_ARM_NONSECURE_FIRMWARE)
```
Also, I am wondering: why is this not becoming part of MDK? They already have a define for TRUSTZONE...

galak commented:
> So @anangl @carlescufi in principle I agree with this change, as long as we guarantee that the DT_* generated defines for peripheral base addresses won't be used in Zephyr, and we will exclusively use the MDK defines, given here.
> 
> Otherwise, it seems we will be creating confusion with the way we extract peripheral addresses

I don't agree, most vendor HALs have some way of defining peripheral addresses and we dont always use them in drivers as we utilize DT based solutions.  One reason maybe to utilize DT defines is if/when we do more code generation its easier to use the DT defines to handle multiple instances of a peripheral in a templated manor.

ioannisg commented:
> 
> 
> > So @anangl @carlescufi in principle I agree with this change, as long as we guarantee that the DT_* generated defines for peripheral base addresses won't be used in Zephyr, and we will exclusively use the MDK defines, given here.
> > Otherwise, it seems we will be creating confusion with the way we extract peripheral addresses
> 
> I don't agree, most vendor HALs have some way of defining peripheral addresses and we dont always use them in drivers as we utilize DT based solutions. One reason maybe to utilize DT defines is if/when we do more code generation its easier to use the DT defines to handle multiple instances of a peripheral in a templated manor.

Well, I see this. My point is, basically: we need to have a **single** way of using nRF peripheral device addresses in Zephyr, preferably with DTS, but, if not, with MDK; I am ok with that. But not both ways.

ioannisg commented on nrfx_config_nrf9160.h:
```diff
@@ -36,243 +36,83 @@
-#define NRF_RTC1 \
-        ((NRF_RTC_Type *)DT_NORDIC_NRF_RTC_RTC_1_BASE_ADDRESS)
-#endif
+#if defined(CONFIG_ARM_NONSECURE_FIRMWARE)
```
By the way @anangl , could this be moved one level up? This is not strictly related to nRF9160.

anangl replied:
You mean to move it to _nrfx_config.h_? Yes, I think it could be moved to make this more future-proof.

ioannisg replied:
Yes, I think so (commented also my last comment)

anangl commented:
@ioannisg 
> So @anangl @carlescufi in principle I agree with this change, as long as we guarantee that the DT_* generated defines for peripheral base addresses won't be used in Zephyr, and we will exclusively use the MDK defines, given here.
> 
> Otherwise, it seems we will be creating confusion with the way we extract peripheral addresses.

Currently nRF9160 is the only nRF chip supported in Zephyr for which we define these `NRF_*` symbols that are needed in nrfx with base addresses taken from devicetree. So this PR actually clean things up by making the way we utilize nrfx consistent.

The only nRF Zephyr drivers that use `DT_*_BASE_ADDRESS` macros are currently the UART ones: [uart_nrfx_uart.c](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/serial/uart_nrfx_uart.c#L17) and [uart_nrfx_uarte.c](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/serial/uart_nrfx_uarte.c#L1290). And they can be easily switched to be using `NRF_*` symbols instead.

We could additionally add a header file that would be included after `<nrf.h>` and that would check at compile time that the base addresses from  devicetree match the ones from MDK. I mean something like this:
```
#if defined(CONFIG_ARM_NONSECURE_FIRMWARE)
#define NRF_BASE_ADDR(P)   P##_NS_BASE
#elif defined(CONFIG_ARMV8_M_SE)
#define NRF_BASE_ADDR(P)   P##_S_BASE
#else
#define NRF_BASE_ADDR(P)   P##_BASE
#endif

#if defined(DT_INST_0_NORDIC_NRF_CLOCK_BASE_ADDRESS) \
        && (DT_INST_0_NORDIC_NRF_CLOCK_BASE_ADDRESS != NRF_BASE_ADDR(NRF_CLOCK))
#error
#endif
...
```
What you think?

ioannisg commented on nrfx_config_nrf9160.h:
```diff
@@ -36,243 +36,83 @@
-#define NRF_RTC1 \
-        ((NRF_RTC_Type *)DT_NORDIC_NRF_RTC_RTC_1_BASE_ADDRESS)
-#endif
+#if defined(CONFIG_ARM_NONSECURE_FIRMWARE)
```
Is it possible to use here the MDK macro for Non-Secure firmware, `NRF_TRUSTZONE_NONSECURE`? Then we will be consistent throughout nRFx source tree.

anangl replied:
Yes, that makes sense.

ioannisg commented:
> 
> 
> @ioannisg
> 
> > So @anangl @carlescufi in principle I agree with this change, as long as we guarantee that the DT_* generated defines for peripheral base addresses won't be used in Zephyr, and we will exclusively use the MDK defines, given here.
> > Otherwise, it seems we will be creating confusion with the way we extract peripheral addresses.
> 
> Currently nRF9160 is the only nRF chip supported in Zephyr for which we define these `NRF_*` symbols that are needed in nrfx with base addresses taken from devicetree. So this PR actually clean things up by making the way we utilize nrfx consistent.
> 
> The only nRF Zephyr drivers that use `DT_*_BASE_ADDRESS` macros are currently the UART ones: [uart_nrfx_uart.c](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/serial/uart_nrfx_uart.c#L17) and [uart_nrfx_uarte.c](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/serial/uart_nrfx_uarte.c#L1290). And they can be easily switched to be using `NRF_*` symbols instead.
> 
> We could additionally add a header file that would be included after `<nrf.h>` and that would check at compile time that the base addresses from devicetree match the ones from MDK. I mean something like this:
> 
> ```
> #if defined(CONFIG_ARM_NONSECURE_FIRMWARE)
> #define NRF_BASE_ADDR(P)   P##_NS_BASE
> #elif defined(CONFIG_ARMV8_M_SE)
> #define NRF_BASE_ADDR(P)   P##_S_BASE
> #else
> #define NRF_BASE_ADDR(P)   P##_BASE
> #endif
> 
> #if defined(DT_INST_0_NORDIC_NRF_CLOCK_BASE_ADDRESS) \
>         && (DT_INST_0_NORDIC_NRF_CLOCK_BASE_ADDRESS != NRF_BASE_ADDR(NRF_CLOCK))
> #error
> #endif
> ...
> ```
> 
> What you think?

Yes, so the keyword here is "currently" :)
So, as I 've said already, I am fine with using MDK instead of DTS for peripheral addresses _also_ for nRF9160, as long as we do it consistently so, yes:
- we need to make the UART drivers switch to using the MDK macros
- we need to have these build-time ASSERTS because otherwise we are prone to errors that are only detected in run-time, e..g with weird BusFaults
- we should move the definition that @carlescufi suggested to one level up so we won't need to duplicate it for future nRF91* variants with TrustZone support

anangl commented:
@ioannisg 

> So, as I 've said already, I am fine with using MDK instead of DTS for peripheral addresses _also_ for nRF9160, as long as we do it consistently so, yes:
> 
> * we need to make the UART drivers switch to using the MDK macros
> * we need to have these build-time ASSERTS because otherwise we are prone to errors that are only detected in run-time, e..g with weird BusFaults
> * we should move the definition that @carlescufi suggested to one level up so we won't need to duplicate it for future nRF91* variants with TrustZone support

I addressed your comment regarding `NRF_TRUSTZONE_NONSECURE` and the last of the above points. Do you agree with addressing the other two in a separate PR in the main Zephyr repo (not in #19269, which would apply only the changes done to this module)?

ioannisg commented:
> 
> 
> @ioannisg
> 
> > So, as I 've said already, I am fine with using MDK instead of DTS for peripheral addresses _also_ for nRF9160, as long as we do it consistently so, yes:
> > 
> > * we need to make the UART drivers switch to using the MDK macros
> > * we need to have these build-time ASSERTS because otherwise we are prone to errors that are only detected in run-time, e..g with weird BusFaults
> > * we should move the definition that @carlescufi suggested to one level up so we won't need to duplicate it for future nRF91* variants with TrustZone support
> 
> I addressed your comment regarding `NRF_TRUSTZONE_NONSECURE` and the last of the above points. Do you agree with addressing the other two in a separate PR in the main Zephyr repo (not in #19269, which would apply only the changes done to this module)?

Yes sure, thanks.

anangl commented:
> Looks good now. Should we run CI in Zephyr with the head of this PR, to make sure things work?

Definitely we should. I waited with this until we reach the agreement here, to not involve CI resources unnecessarily.

ioannisg commented:
> 
> 
> > Looks good now. Should we run CI in Zephyr with the head of this PR, to make sure things work?
> 
> Definitely we should. I waited with this until we reach the agreement here, to not involve CI resources unnecessarily.

OK, let me know how CI went ;)

anangl commented:
> OK, let me know how CI went ;)

It went [green](https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/51288/summary), so I guess we're ready to merge.

ioannisg commented:
> 
> 
> > OK, let me know how CI went ;)
> 
> It went [green](https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/51288/summary), so I guess we're ready to merge.

Great. How about the build-time checks for DT_ macros, could you send this patch as well?

anangl commented:
> How about the build-time checks for DT_ macros, could you send this patch as well?

Sure. But not today, after the weekend.
